File: app/models/system.py
from typing import List, Optional

from pydantic import BaseModel, Field


class UsageStats(BaseModel):
    current: int
    total: int
    percent: float


class HistoryEntry(BaseModel):
    timestamp: int
    value: float


class NetworkHistoryEntry(BaseModel):
    timestamp: int
    incoming: int
    outgoing: int


class PersonalUsageStats(BaseModel):
    total_users: int
    consumed_bytes: int
    built_bytes: int
    reset_bytes: int


class AdminOverviewStats(BaseModel):
    total_admins: int
    sudo_admins: int
    full_access_admins: int
    standard_admins: int
    top_admin_username: Optional[str] = None
    top_admin_usage: int = 0


class RedisStats(BaseModel):
    enabled: bool
    connected: bool
    memory_used: int = 0
    memory_total: int = 0
    memory_percent: float = 0.0
    uptime_seconds: int = 0
    version: Optional[str] = None
    keys_count: int = 0
    keys_cached: int = 0
    commands_processed: int = 0
    hits: int = 0
    misses: int = 0
    hit_rate: float = 0.0


class SystemStats(BaseModel):
    version: str
    cpu_cores: int
    cpu_usage: float
    total_user: int
    online_users: int
    users_active: int
    users_on_hold: int
    users_disabled: int
    users_expired: int
    users_limited: int
    incoming_bandwidth: int
    outgoing_bandwidth: int
    panel_total_bandwidth: int
    incoming_bandwidth_speed: int
    outgoing_bandwidth_speed: int
    memory: UsageStats
    swap: UsageStats
    disk: UsageStats
    load_avg: List[float] = Field(default_factory=list)
    uptime_seconds: int
    panel_uptime_seconds: int
    xray_uptime_seconds: int
    xray_running: bool
    xray_version: Optional[str] = None
    app_memory: int
    app_threads: int
    panel_cpu_percent: float
    panel_memory_percent: float
    cpu_history: List[HistoryEntry] = Field(default_factory=list)
    memory_history: List[HistoryEntry] = Field(default_factory=list)
    network_history: List[NetworkHistoryEntry] = Field(default_factory=list)
    panel_cpu_history: List[HistoryEntry] = Field(default_factory=list)
    panel_memory_history: List[HistoryEntry] = Field(default_factory=list)
    personal_usage: PersonalUsageStats
    admin_overview: AdminOverviewStats
    last_xray_error: Optional[str] = None
    redis_stats: Optional[RedisStats] = None

File: dashboard/src/types/System.ts
export interface UsageStats {
	current: number;
	total: number;
	percent: number;
}

export interface HistoryEntry {
	timestamp: number;
	value: number;
}

export interface NetworkHistoryEntry {
	timestamp: number;
	incoming: number;
	outgoing: number;
}

export interface PersonalUsageStats {
	total_users: number;
	consumed_bytes: number;
	built_bytes: number;
	reset_bytes: number;
}

export interface AdminOverviewStats {
	total_admins: number;
	sudo_admins: number;
	full_access_admins: number;
	standard_admins: number;
	top_admin_username?: string | null;
	top_admin_usage: number;
}

export interface RedisStats {
	enabled: boolean;
	connected: boolean;
	memory_used: number;
	memory_total: number;
	memory_percent: number;
	uptime_seconds: number;
	version?: string | null;
	keys_count: number;
	keys_cached: number;
	commands_processed: number;
	hits: number;
	misses: number;
	hit_rate: number;
}

export interface SystemStats {
	version: string;
	cpu_cores: number;
	cpu_usage: number;
	total_user: number;
	online_users: number;
	users_active: number;
	users_on_hold: number;
	users_disabled: number;
	users_expired: number;
	users_limited: number;
	incoming_bandwidth: number;
	outgoing_bandwidth: number;
	panel_total_bandwidth: number;
	incoming_bandwidth_speed: number;
	outgoing_bandwidth_speed: number;
	memory: UsageStats;
	swap: UsageStats;
	disk: UsageStats;
	load_avg: number[];
	uptime_seconds: number;
	panel_uptime_seconds: number;
	xray_uptime_seconds: number;
	xray_running: boolean;
	xray_version?: string | null;
	app_memory: number;
	app_threads: number;
	panel_cpu_percent: number;
	panel_memory_percent: number;
	cpu_history: HistoryEntry[];
	memory_history: HistoryEntry[];
	network_history: NetworkHistoryEntry[];
	panel_cpu_history: HistoryEntry[];
	panel_memory_history: HistoryEntry[];
	personal_usage: PersonalUsageStats;
	admin_overview: AdminOverviewStats;
	last_xray_error?: string | null;
	last_telegram_error?: string | null;
	redis_stats?: RedisStats | null;
}
